pasterd.8
.\"
.\" Copyright (c) 2020-2023 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd 06 February, 2020
.Dt PASTERD 8
.Os
.\" NAME
.Sh NAME
.Nm pasterd
.Nd simple paste service
.\" SYNOPSIS
.Sh SYNOPSIS
.Nm
.Op Fl qv
.Op Fl d Ar database-path
.Op Fl t Ar theme-directory
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
utility is a simple FastCGI program to host code snippets over a web interface.
It will show most recent public pastes and let users to create new one from a
web form.
.Pp
It supports:
.Bl -bullet -compat
.It
Listing of recent pastes,
.It
Submission of new pastes,
.It
Searching existing pastes,
.It
Language highlighting (depending on the theme),
.It
Private pastes (not listed).
.El
.Pp
To store pastes,
.Nm
uses a SQLite database that must be writable by the CGI/FastCGI owner. See usage
below.
.Pp
Available options:
.Bl -tag -width Ds
.It Fl d Ar database-path
Specify an alternate path for the database.
.It Fl t Ar theme-directory
Specify an alternate directory for the theme.
.It Fl q
Do not log through syslog at all.
.It Fl v
Increase verbosity level.
.El
.\" USAGE
.Sh USAGE
The
.Nm
utility does not use configuration file as it does not need many adjustments,
instead every parameter could be passed by environment variables or options.
.Pp
By default,
.Nm
will try to use
.Pa @VARDIR@/paster/paster.db
database.
.\" LOGS
.Sh LOGS
The
.Nm
utility will log information through syslog unless verbosity is disabled.
Except at startup where the tool can write to stderr some information if it
can't continue processing, the tool will never write anything to stdout and
use syslog only.
.Pp
The available verbosity level is defined in the following order:
.Bd -literal -offset Ds
none < warnings (default) < info < debug
.Ed
.Pp
Use
.Fl q
or
.Va PASTERD_VERBOSITY=0
if you want to disable syslog completely.
.\" USING WITH FASTCGI
.Sh USING WITH FASTCGI
The recommended way to use
.Nm
is to deploy using FastCGI. You can use the
.Xr kfcgi 8
helper to spawn the process for you.
.Pp
Example:
.Bd -literal -offset Ds
kfcgi -p /var/www/paster -- pasterd -f -d paster.db -t siimple
.Ed
.Pp
Note: kfcgi chroot to the directory given, you must either statically link
pasterd at build time or deploy all required libraries. Also, themes directory
will need to be available in the chroot directory. In the above example, this
will effectively create a database
.Pa /var/www/paster/paster.db
and use the theme
.Pa /var/www/paster/siimple .
.Pp
Then, simply copy the desired theme into the directory.
.Bd -literal -offset Ds
cp -R @SHAREDIR@/paster/themes/siimple /var/www/paster
.Ed
.Pp
As an
.Em insecure
alternative, you can chroot to
.Pa /
to avoid static-linking and copying themes, using:
.Bd -literal -offset Ds
kfcgi -p / -- pasterd -f \e
	-d /var/www/paster/paster.db \e
	-t @SHAREDIR@/paster/themes/siimple
.Ed
.Pp
Both kfcgi invocations will create
.Pa /var/www/run/http.sock
with current user and group. Configure the web server to talk to that socket
and make sure it has appropriate file permissions otherwise see
.Fl u
option in
.Nm kfcgi .
See also the
.Xr kfcgi 8
manual for more information.
.Pp
Next, configure the web server.
.Pp
Warning: at this moment,
.Nm
requires its own virtual host and can
.Em not
use a url.
.\" Server: nginx
.Ss Server: nginx
The nginx web server requires several parameters to run
.Nm .
.Bd -literal
server {
	server_name mypaste.fr;
	listen 80;

	location / {
		fastcgi_param QUERY_STRING      query_string;
		fastcgi_param REQUEST_METHOD    $request_method;
		fastcgi_param CONTENT_TYPE      $content_type;
		fastcgi_param CONTENT_LENGTH    $content_length;
		fastcgi_param SCRIPT_FILENAME   $document_root$fastcgi_script_name;
		fastcgi_param SCRIPT_NAME       $fastcgi_script_name;
		fastcgi_param PATH_INFO         $document_uri;
		fastcgi_param PATH_TRANSLATED   $document_root$fastcgi_path_info;
		fastcgi_param REQUEST_URI       $request_uri;
		fastcgi_param DOCUMENT_URI      $document_uri;
		fastcgi_param DOCUMENT_ROOT     $document_root;
		fastcgi_param SERVER_PROTOCOL   $server_protocol;
		fastcgi_param GATEWAY_INTERFACE CGI/1.1;
		fastcgi_param SERVER_SOFTWARE   nginx/$nginx_version;
		fastcgi_param REMOTE_ADDR       $remote_addr;
		fastcgi_param REMOTE_PORT       $remote_port;
		fastcgi_param SERVER_ADDR       $server_addr;
		fastcgi_param SERVER_PORT       $server_port;
		fastcgi_param SERVER_NAME       $server_name;
		fastcgi_param HTTPS             $https;
		fastcgi_pass unix:/var/www/run/httpd.sock;
	}
}
.Ed
.\" ENVIRONMENT
.Sh ENVIRONMENT
The following environment variables are detected:
.Bl -tag -width Ds
.It Va PASTERD_DATABASE_PATH No (string)
Path to the SQLite database.
.It Va PASTERD_THEME_DIR No (string)
Directory containing the theme.
.It Va PASTERD_VERBOSITY No (number)
Verbosity level, 0 to disable completely.
.El
.\" AUTHORS
.Sh AUTHORS
.Nm
was written by David Demelier <markand@malikania.fr>
.\" SEE ALSO
.Sh SEE ALSO
.Xr paster 8 ,
.Xr kfcgi 8


CHANGES.md
paster CHANGES
==============

paster -- current
-----------------------

- Pastes are private by default with paster(1) (#2481).

paster 0.2.1 2020-02-14
-----------------------

- Fix several database issues.

paster 0.2.0 2020-02-13
-----------------------

- Fix database lock error (#2475),
- Added unit tests for database (#1696).

paster 0.1.0 2020-02-07
-----------------------

- Initial release.


pasterd.c
/*
 * pasterd.c -- main pasterd(8) file
 *
 * Copyright (c) 2020-2023 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "config.h"
#include "database.h"
#include "http.h"
#include "log.h"
#include "util.h"

/*
 * Interval in seconds for each cleanup time. Since a paste has one hour
 * duration as minimal let's cleanup every hour.
 */
#define CLEANUP_INTERVAL 3600

static timer_t timer;

/*
 * This function runs in a thread, we open our own local database to let the
 * engine locks by itself.
 */
static void
cleanup(union sigval val)
{
	(void)val;

	struct database db;

	if (database_open(&db, config.databasepath) == 0) {
		database_clear(&db);
		database_finish(&db);
	}
}

static void
defaults(void)
{
	snprintf(config.databasepath, sizeof (config.databasepath),
	    "%s", VARDIR "/paster/paster.db");
}

static void
init(void)
{
	struct sigevent sigev = {
		.sigev_notify = SIGEV_THREAD,
		.sigev_notify_function = cleanup
	};
	struct itimerspec spec = {
		.it_value = { .tv_sec = CLEANUP_INTERVAL },
		.it_interval = { .tv_sec = CLEANUP_INTERVAL }
	};

	srand(time(NULL));
	log_open();

	if (!config.databasepath[0])
		die("abort: no database specified\n");
	if (database_open(&database, config.databasepath) < 0)
		die("abort: could not open database\n");
	if (timer_create(CLOCK_MONOTONIC, &sigev, &timer) < 0)
		die("abort: timer_create: %s\n", strerror(errno));
	if (timer_settime(timer, 0, &spec, NULL) < 0)
		die("abort: timer_settime: %s\n", strerror(errno));
}

static void
run(void)
{
	http_fcgi_run();
}

static void
finish(void)
{
	database_finish(&database);
	log_finish();
}

static void
usage(void)
{
	fprintf(stderr, "usage: paster [-qv] [-d database-path] [-t theme-directory]\n");
	exit(1);
}

int
main(int argc, char **argv)
{
	const char *value;
	int opt;

	defaults();

	/* Seek environment variables before options. */
	if ((value = getenv("PASTERD_DATABASE_PATH")))
		snprintf(config.databasepath, sizeof (config.databasepath), "%s", value);
	if ((value = getenv("PASTERD_THEME_DIR")))
		snprintf(config.themedir, sizeof (config.themedir), "%s", value);
	if ((value = getenv("PASTERD_VERBOSITY")))
		config.verbosity = atoi(value);

	while ((opt = getopt(argc, argv, "d:t:qv")) != -1) {
		switch (opt) {
		case 'd':
			snprintf(config.databasepath, sizeof (config.databasepath), "%s", optarg);
			break;
		case 't':
			snprintf(config.themedir, sizeof (config.themedir), "%s", optarg);
			break;
		case 'v':
			config.verbosity++;
			break;
		case 'q':
			config.verbosity = 0;
			break;
		default:
			usage();
			break;
		}
	}

	init();
	run();
	finish();
}
